lib/macroni/Conversion/Kernel/KernelRewriters.cpp

```diff
@@ -1,254 +1,235 @@
+  exp.getExpansion().walk([&](mlir::Operation *op) {
+    if (auto param = mlir::dyn_cast<macroni::MacroParameter>(op)) {
+      auto param_name = param.getParameterName();
+      if (param_name == "p") {
+        p = op;
+      }
     }
-
-    mlir::LogicalResult rewrite_container_of(
-        macroni::MacroExpansion exp,
-        mlir::PatternRewriter &rewriter) {
-        if (!is_container_of(exp)) {
-            return mlir::failure();
-        }
-        mlir::Operation *ptr = nullptr;
-        std::optional<mlir::TypeAttr> type;
-        std::optional<mlir::StringAttr> member;
-        exp.getExpansion().walk([&](mlir::Operation *op) {
-            if (auto mp = mlir::dyn_cast<macroni::MacroParameter>(op)) {
-                auto param_name = mp.getParameterName();
-                if ("ptr" == param_name) {
-                    ptr = op;
-                }
-            } else if (auto member_op
-                       = mlir::dyn_cast<vast::hl::RecordMemberOp>(op)) {
-                auto record = member_op.getRecord();
-                auto record_ty = record.getType();
-                if (auto pty =
-                    mlir::dyn_cast<vast::hl::PointerType>(record_ty)) {
-                    auto elem_ty = pty.getElementType();
-                    auto ty_attr = mlir::TypeAttr::get(elem_ty);
-                    type = ty_attr;
-                } else {
-                    auto ty_attr = mlir::TypeAttr::get(record_ty);
-                    type = ty_attr;
-                }
-                member = member_op.getNameAttr();
-            }});
-        if (!(ptr && type && member && ptr->getNumResults() == 1)) {
-            return mlir::failure();
-        }
-
-        auto ptr_clone = rewriter.clone(*ptr);
-        rewriter.replaceOpWithNewOp<ContainerOf>(exp, exp.getType(0),
-                                                 ptr_clone->getResult(0), *type,
-                                                 *member);
-        return mlir::success();
+  });
+  if (!(p && p->getNumResults() == 1)) {
+    return mlir::failure();
+  }
+
+  auto p_clone = rewriter.clone(*p);
+  rewriter.replaceOpWithNewOp<RCUDereference>(exp, exp.getType(0),
+                                              p_clone->getResult(0));
+  return mlir::success();
+}
+
+mlir::LogicalResult rewrite_smp_mb(macroni::MacroExpansion exp,
+                                   mlir::PatternRewriter &rewriter) {
+  if (!is_smp_mb(exp)) {
+    return mlir::failure();
+  }
+
+  rewriter.replaceOpWithNewOp<kernel::SMPMB>(exp, exp.getType(0));
+  return mlir::success();
+}
+
+mlir::LogicalResult rewrite_list_for_each(vast::hl::ForOp for_op,
+                                          mlir::PatternRewriter &rewriter) {
+  mlir::Operation *pos = nullptr;
+  mlir::Operation *head = nullptr;
+  for_op.getCondRegion().walk([&](mlir::Operation *op) {
+    if (auto param_op = mlir::dyn_cast<macroni::MacroParameter>(op)) {
+      if (param_op.getParameterName() == "pos") {
```

Comment:
This is still the same only the string changes -> should be in the code only once then.